### ISSUE TITLE
Azure Go SDK - Support for Multi Tenant Authentication

### ISSUE THREAD

**Issue by skywalker2909** [Azure/azure-sdk-for-go#19726]
### Feature Request

I am trying to implement ```multi tenant``` authentication using the sdk but cannot find any code samples or godocs for the same.

I came across this Microsoft devblog post on [Guidance for multi-tenant applications using the Azure Identity libraries](https://devblogs.microsoft.com/azure-sdk/guidance-for-multi-tenant-applications-using-the-azure-identity-libraries/#additional-recommended-actions) but this does not mention any example for golang.

I am looking to implement something similar using the go sdk like this .NET example

```
var credential = new DefaultAzureCredential(new DefaultAzureCredentialOptions
{
    AdditionallyAllowedTenants = { "<tenant_id_1>", "<tenant_id_2>" }
});
```
I checked for documentation on in it in the ```azidentity``` go package docs but couldn't find anything relevant.

Does the go sdk support this? How do i go about implementing it?

**Comment by chlowell** [comment]
Thanks for opening this. The `AdditionallyAllowedTenants` API is part of #17159 but we didn't have an issue tracking it specifically; let's use this one. We plan to release a beta with this API in early 2023. Usage will look like this:
```go
// cred will request tokens from the configured tenant unless another is specified in GetToken args
cred, err := azidentity.NewClientSecretCredential("primary-tenant", "client-id", "secret", &azidentity.ClientSecretCredentialOptions{
	// cred will (attempt to) authenticate in any requested tenant; can also list specific tenants here
	AdditionallyAllowedTenants: []string{"*"},
})
```

**Comment by skywalker2909** [comment]
Thanks for getting back @chlowell.

Is there any other workaround for this in golang if i want to implement this now?

**Comment by chlowell** [comment]
Yes, thanks for asking! I should have suggested one. Today, a specific tenant is baked into most credentials and thus into clients as well, so you must create one of each for every tenant your application accesses. You could consolidate that pattern into a helper like this:
```go
func GetClientForTenant(tenant, clientID, secret string) (*armsubscription.Client, error) {
	cred, err := azidentity.NewClientSecretCredential(tenant, clientID, secret, nil)
	if err != nil {
		return nil, err
	}
	return armsubscription.NewSubscriptionsClient(cred, nil)
}
```
(adjust for the credential and client types your application uses)

**Comment by skywalker2909** [comment]
Thanks @chlowell 

Although i doubt i will be able to use the solution you provided for authenticating across tenants in different azure organisations?

Something similar to how its mentioned in the docs here - [Authenticate requests across tenants](https://learn.microsoft.com/en-us/azure/azure-resource-manager/management/authenticate-multi-tenant)

Like for example, if a single sdk call involves making changes in 2 different azure tenants residing in different organisations (VNet peering for example), then i would need to pass in an ```x-ms-authorization-auxiliary``` token along with a ```Bearer token```

Any idea how do i go about implementing this approach?

**Comment by chlowell** [comment]
My example above shows how to work around the single tenant nature of `azidentity`. You're correct that this is only part of what you need for that ARM feature. While it is possible for you to implement the rest in your application today, doing so would be complex and I think not worth the effort, unless you really urgently need the feature. I plan to ship Azure SDK libraries with beta support for ARM cross-tenant auth soon (tracked by #17159).

**Comment by skywalker2909** [comment]
Thanks @chlowell for the detailed example. I will implement this for my solution till the PR gets merged.

Just one question - The ```Do()``` method would be called internally by the sdk when the ARM client is instantiated?

**Comment by chlowell** [comment]
The client will call `Do()` for every request i.e., every request it sends to ARM will carry one token from each credential.

**Comment by skywalker2909** [comment]
Got it. 

Thank you @chlowell for all your help and support on this. Much appreciated!
